Extension for running Narayana LRA participants

mmusgrov opened the pull request:
#4399 

This PR adds support for an implementation of the eclipse microprofile-lra project version 1.0 (https://github.com/eclipse/microprofile-lra/releases/tag/1.0). I want to target the extension to core because compensating transactions are an essential component for building reliable services (and in this respect is similar/complementary to the narayana-jta extension which itself is useful for doing each of the compensatory actions that comprise the LRA).

It replaces PR #6763 which was closed as out of date. I was unable to re-open it but the discussion provides useful background and context for this PR.

mmusgrov committed:
issue4399 Extension for running Narayana LRA participants

NahuelOlgiati commented on extensions/narayana-lra/deployment/src/main/java/io/quarkus/narayana/lra/deployment/NarayanaLRAProcessor.java:
```diff
@@ -0,0 +1,37 @@
+import io.quarkus.narayana.lra.runtime.LRAConfiguration;
+import io.quarkus.narayana.lra.runtime.NarayanaLRAProducers;
+import io.quarkus.narayana.lra.runtime.NarayanaLRARecorder;
+
```
I think this class should set the Feature. Like feature.produce(new FeatureBuildItem(Feature.NARAYANA_LRA));

mmusgrov replied:
okay will do

cescoffier replied:
Seems to be done in the registerFeature method.

mmusgrov replied:
Yes, I made changes after the first round of reviews.

NahuelOlgiati commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+
+[source,bash]
+----
+quarkus.lra.coordinator-url=http://localhost:8080/lra-coordinator
```
Maybe "quarkus.narayana.lra.coordinator-url" is better than "quarkus.lra.coordinator-url"

mmusgrov replied:
Makes sense, I will add this change too.

But I will wait a while for anyone else who wants to review the PR before pushing a commit (since the PR check took about 5 hours last time). I will give it a few days and then ping someone from the quarkus team to check whether or not your review is sufficient to get the PR merged.

mmusgrov replied:
So it's a land grab, I have no issues with that!

mmusgrov replied:
So on a second reading I am going to keep the property name as it was (namely, `quarkus.lra.coordinator-url`). - it's how the quarkus config property naming system is meant to work.

mmusgrov replied:
@nahuelolgiati (Nahuel Olgiati) this was your only review comment that I declined. If that's okay and if the rest of the updates I made to the PR are okay too then if you have the permissions, will you mark it as approved.

mmusgrov commented:
@quarkusio/committerhelp I have had one review so far, what is the process for getting things merged once I have addressed all reviewer comments. Note that I am still waiting for the original reviewer (@xstefank ) to provide his review.

xstefank commented on integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRAParticipantTestResourceLifecycle.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public Map<String, String> start() {
+        registry = new GenericContainer<>("jbosstm/lra-coordinator:5.12.0.Final")
```
Maybe we can use latest here or parse the version dynamically from the properties? More a question to the quarkus team in how this is handled elsewhere.

mmusgrov replied:
It's a test. And other integration tests use a similar approach.

xstefank replied:
ok, just to remember to bump this version when the Narayana update will be done all the time seems a little overkill. Not sure if dependabot can handle such a scenario.

mmusgrov replied:
When we migrate to a new version of narayana the version will need updating in various places, I'm still not seeing the benefit of making this particular change generic (if I could even figure out a clean way to do that).

cescoffier replied:
dependabot won't update it.
But, I believe it's fine.

mmusgrov replied:
okay

mmusgrov replied:
What is a periodic reminder?

cescoffier replied:
a periodic todo task that remind you to update the version once in a while.

mmusgrov replied:
I am not sure what you are asking us to do but we try to track our users and update accordingly, for example we already do this with wildfly and quarkus is another platform we monitor,

xstefank commented on integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRAParticipantTestResourceLifecycle.java:
```diff
@@ -0,0 +1,55 @@
+    public Map<String, String> start() {
+        registry = new GenericContainer<>("jbosstm/lra-coordinator:5.12.0.Final")
+                //.withNetworkMode("host"); // fails with: "Container doesn't expose any ports"
+                //.withExposedPorts(90010); // and exposing one then fails with: "Requested port (90010) is not mapped"
```
Not sure if this has value, do you think that others will run into similar issues?

mmusgrov replied:
It was an attempt at working around a resteasy integration issue about which I still need to talk to the resteasy integration team . But, again, if you think it's going to be an issue I can remove it. On the hand leaving it in will help other developers who hit the same issue.

xstefank replied:
maybe you can open a new issue if this needs to be fixed elsewhere where you will reference a branch in your fork with the left comments to be fixed and then you can open a follow-up PR updating to the correct configuration. So please remove the unused code.

mmusgrov replied:
I don't think so. But if you insist then I will remove them.

cescoffier replied:
The only exposed port is 8080/tcp. So you would need .withExposedPorts(8080) and retrieved the mapped port.

mmusgrov replied:
I already to that, if you check the code in the PR you can see that we are already exposing 8080.

mmusgrov commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+
+== Introduction
+
+The LRA (short for Long Running Action) participant extension is useful in microservice based
```
An LRA is a long running action.

xstefank replied:
ok, I think the specs is plural https://github.com/eclipse/microprofile-lra#long-running-actions-for-microprofile but I have no issue with this.

mmusgrov replied:
A participant participates in a Long Running Action.

mmusgrov commented on core/deployment/src/main/java/io/quarkus/deployment/Capability.java:
```diff
@@ -29,6 +29,8 @@
      */
     String TRANSACTIONS = QUARKUS_PREFIX + "transactions";
 
+    String LRA = QUARKUS_PREFIX + "lra.participant";
```
okay

mmusgrov replied:
But the capability is participant support for lra so "lra.participant" is the natural choice.

mmusgrov replied:
Ah yes, its the name of the constant that was the issue, sorry I was focusing in on the value stored in the constant.

mmusgrov commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+quarkus.lra.coordinator-url=http://localhost:8080/lra-coordinator
+----
+
+For a narayana coordinator the path component of the url is normally `lra-coordinator`.
```
okay

mmusgrov replied:
On a second reading, capitalising it is a bit shouty, I don't see any problem with lower case.

mmusgrov commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+the coordinator or for repairing the network, respectively). To fulfil this task the
+coordinator must have access to durable storage for its logs (via a filesystem or in
+a database).  At the time of writing, managing coordinators is the responsibility of
+the user. An "out-of-the-box" solution will be forthcoming.
```
Because the user has to manage it.

Example "out-of-the-box" solutions will be WildFly and an OCP operator.

xstefank replied:
ok

mmusgrov commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+the method finishes and the `end = true` element on the confirmTrip method forces the LRA
+(started by the bookTrip method) to close the LRA. Note that this end element can
+be placed on any JAX-RS resource (ie one service can start the LRA whilst a different
+service ends it).
```
No, the guide is to help users get started with the extension.
There is only one configuration option and it does not need a specific section for it. The text that refers to the option flows just fine with the section where I placed it.

xstefank replied:
ok

cescoffier replied:
It's very interesting, but we may want to refactor this document into:

1. Getting started with LRA - just plain getting started, nothing else 
2. An LRA reference guide with these examples

All this content is great, but we also need to distillate it at the right pace for users that are not in the LRA world. It's a difficult exercise, and I/we failed multiple times.

mmusgrov replied:
So when there is some guidance I will split it up.
I intentionally made the guide minimal since there is content elsewhere and I was keen to avoid duplication and the maintenance impact that brings.

cescoffier replied:
Here are some rules I'm trying to apply:

* getting started should only contain details to run an example with a small introduction. It should not expect any prior knowledge, and from the beginning to the end should not take more than 15 minutes. 
* reference contains everything else, more complex example, configuration.... 

So, typically, from line 164, it should be in the reference guide. 
Note that of course, you can cross-link.

mmusgrov replied:
The guide is already minimal, so you are requesting that I reduce it even more?

cescoffier replied:
The last part looks more advanced. At least it's how I felt it when I read it.

mmusgrov replied:
We are doing microservices.

mmusgrov commented on extensions/narayana-lra/deployment/src/main/java/io/quarkus/narayana/lra/deployment/NarayanaLRAProcessor.java:
```diff
@@ -0,0 +1,37 @@
+
+    @BuildStep
+    BeanDefiningAnnotationBuildItem additionalBeanDefiningAnnotation() {
+        return new BeanDefiningAnnotationBuildItem(DotName.createSimple(LRAParticipantRegistry.class.getName()));
```
This extension was originally developed over a year and a half ago so it may be a carry over from one of the earlier versions.
I will check it.

mmusgrov commented on extensions/narayana-lra/runtime/pom.xml:
```diff
@@ -0,0 +1,86 @@
+
+    <properties>
+        <narayana.version>5.12.0.Final</narayana.version>
+        <version.microprofile.lra.api>1.0</version.microprofile.lra.api>
```
which bom?

xstefank replied:
https://github.com/quarkusio/quarkus/blob/main/bom/application/pom.xml#L37

cescoffier replied:
Yes, all versions should be in the quarkus bom.

mmusgrov commented on extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/NarayanaLRAProducers.java:
```diff
@@ -0,0 +1,14 @@
+public class NarayanaLRAProducers {
+    @Produces
+    public LRAParticipantRegistry lraParticipantRegistry() {
+        return null;
```
I don't think so. Producing a null value is explicitly allowed by [the spec](https://docs.jboss.org/cdi/api/2.0/javax/enterprise/inject/Produces.html).

xstefank replied:
ok

mmusgrov commented on extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/NarayanaLRARecorder.java:
```diff
@@ -0,0 +1,13 @@
+@Recorder
+public class NarayanaLRARecorder {
+    public void setConfig(final LRAConfiguration config) {
+        if (System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_URL_KEY) == null) {
```
Not sure what you mean but the line of code is intentional.

xstefank replied:
users can do `java -jar lra-quarkus.jar -Dquarkus.lra.coordinator.url=url1 -Dlra.coordinator.url=url2` which means that starting LRA with `@LRA` annotation and calling `lraClient.getAllLRAs()` in this method would not return the started LRA. `lra.coordinator.url` is `NarayanaLRAClient.LRA_COORDINATOR_URL_KEY`. So that's why I'm asking if it is intended as it seems a little strange to me. But maybe there is a use case.

mmusgrov replied:
The line of code was intentional. It allows users to set the url on the java command line. But if it is going to be a blocker for this PR then I'd reluctantly remove it from the PR.

xstefank replied:
Not a blocker but it's good to point out that this can happen. Maybe I really just don't see the use case.

cescoffier replied:
So, it can be set from a sys variable, but not an env one?  In addition, it overrides the configured value.  
Is it because the TCK expect it, or what's the reason behind it?

The configured value can be set using -Dquarkus.lra.coordinator-url and from the QUARKUS_LRA_COORDINATOR_URL env variable.

mmusgrov replied:
Both of you complained about this so I guess I will have to remove it.

mmusgrov replied:
I remember now why I needed it. The integration test starts the coordinator in a docker container and the port mapping needs to be exposed to the extension [1]. The javadoc for QuarkusTestResourceLifecycleManager [2] implies that it will set system properties depending on what the start method returns [3]. But this does not appear to work as expected so I set it as a system property [1] and avoid overwriting it in the extension [4].

So @cescoffier , it is required.

[1] https://github.com/mmusgrov/quarkus/blob/issue4399-lra-extension/integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRAParticipantTestResourceLifecycle.java#L38
[2] https://github.com/quarkusio/quarkus/blob/main/test-framework/common/src/main/java/io/quarkus/test/common/QuarkusTestResourceLifecycleManager.java#L28
[3] https://github.com/mmusgrov/quarkus/blob/issue4399-lra-extension/integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRAParticipantTestResourceLifecycle.java#L40
[4] https://github.com/mmusgrov/quarkus/blob/issue4399-lra-extension/extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/NarayanaLRARecorder.java#L9

cescoffier replied:
Doesn't `quarkus.lra.coordinator-url` work? IF your test resource returns a Map containing this property it should configure the extension with it.

mmusgrov replied:
No it doesn't work.

mmusgrov replied:
I mentioned this fact in my comment, quote: `The javadoc for QuarkusTestResourceLifecycleManager [2] implies that it will set system properties depending on what the start method returns [3]`. And the code already does return it.

cescoffier replied:
You need to return a `Properties` object containing the property to set. Do not set SystemPRoperty directly from TestResource.

mmusgrov replied:
To repeat my last comment `I mentioned this fact in my comment, quote: The javadoc for QuarkusTestResourceLifecycleManager [2] implies that it will set system properties depending on what the start method returns [3]. And the code already does return it.` If you look at the code in the PR you will see that this is what I am doing.

mmusgrov commented on integration-tests/narayana-lra/pom.xml:
```diff
@@ -0,0 +1,170 @@
+
+    <properties>
+        <quarkus.http.test-port>8081</quarkus.http.test-port>
+        <narayana.version>5.11.2.Final</narayana.version>
```
Agreed, thanks.

mmusgrov commented on integration-tests/narayana-lra/pom.xml:
```diff
@@ -0,0 +1,170 @@
+    </build>
+    <profiles>
+        <profile>
+            <id>test-elasticsearch</id>
```
Agreed, I'll rename it to something meaningful to LRA

mmusgrov commented on integration-tests/narayana-lra/src/main/java/org/acme/quickstart/lra/coordinator/TransactionalResource.java:
```diff
@@ -0,0 +1,98 @@
+    @Produces(MediaType.TEXT_PLAIN)
+    public String status() {
+        return "hello";
+    }
```
:-) it was a sanity check, I will remove it.

mmusgrov commented on integration-tests/narayana-lra/src/main/java/org/acme/quickstart/lra/coordinator/TransactionalResource.java:
```diff
@@ -0,0 +1,98 @@
+    @PUT
+    @Path("compensate")
+    @Compensate
+    public Response compensateWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId, String userData) {
```
That's a shame, users would find it extremely useful. I will check whether or not it still works in the narayana implementation of the LRA spec and if not then I will remove it.

xstefank replied:
we can have this on top of the spec in the Narayana implementation as an experimental feature, agreed it would be nice to have.

mmusgrov replied:
It's not an experimental feature, it is part of the narayana implementation of LRAs and has been present, I believe (I would need to look through the history) ever since I took the spec to the MP group. Don't forget that the MP-LRA spec process removed a lot of useful stuff from the implementation and  that the narayana LRA implementation is not restricted to what the spec says, it is a transactional solution that just happens to implement all of the requirements of the MP-LRA spec.

xstefank replied:
true, not experimental, additional, not speced. Sorry.

mmusgrov commented on integration-tests/narayana-lra/src/main/java/org/acme/quickstart/lra/coordinator/TransactionalResource.java:
```diff
@@ -0,0 +1,98 @@
+    @PUT
+    @Path("complete")
+    @Complete
+    public Response completeWork(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId, String userData) {
```
Same response.

mmusgrov commented on integration-tests/narayana-lra/src/main/resources/application.properties:
```diff
@@ -0,0 +1,11 @@
+quarkus.http.test-port=8081
+quarkus.http.host=${HOST:localhost}
+#quarkus.http.host=0.0.0.0
+#%test.quarkus.http.host=0.0.0.0
```
Well they should work so I'd left it in so that we don't forget. I still need to talk to the resteasy integration team about the problem. But if you think it's going to be an issue I can remove it.

xstefank replied:
If they are not used I would suggest removing them, we can add them when they add some value but now it only means that anyone reading the code will stop thinking why this is in here and why it is commented.

mmusgrov replied:
You missed my point but I can remove them.

xstefank reviewed (approved):
Thanks, LGTM but someone from the Quarkus team should check if the extension stuff is in the right places.

mmusgrov commented:
If anyone from the quarkusio team is looking at this, the test failures concern io.quarkus.cli and look unrelated to the PR. Note also that the initial CI run passed and the second run included minor changes only after requests by the reviewers.

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+The approach is for the developer to annotate a business method with a Java annotation
+(https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java[`@LRA`]).
+When such a method is called a context is created (if one is not already present) which is passed
+along with subsequent JAX-RS invocations until a method is reached
```
More for my understanding, is it only JAX-RS? What about gRPC and other protocols?

mmusgrov replied:
Although narayana supports POJO based participants, this version of the extension only supports JAX-RS participants. The reason for this is that our narayana implementation [of MP-LRA] requires a CDI extension but these are not supported by quarkus.

If we cannot workaround this missing feature then we'll update narayana by removing the reliance on CDI extensions. Note that our initial implementation (of LRA) did not require CDI and we only added it for expediency, I think, i.e. it will be a simple exercise to strip out the CDI extension.

FYI: So the next version of the extension will support non JAX-RS participants where the participant only needs the annotations. The current narayana implementation uses a (JAX-RS) proxy to manage the interactions with the coordinator.

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+(https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java[`@LRA`]).
+When such a method is called a context is created (if one is not already present) which is passed
+along with subsequent JAX-RS invocations until a method is reached
+which also contains an `@LRA` annotation with an attribute that indicates that the LRA should be
```
I may have to look at the spec, but I'm confused between `@LRA` and LRA participants. I believe these terms need to be introduced.

mmusgrov replied:
So you got your answer in the next paragraph. Do you think I need to do some reordering of the text where I introduce the various concepts?

mmusgrov replied:
But I will include it here for clarity:

@LRA is the annotation for controlling the [life cycle of a Long Running Action](https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java#L185).

An LRA participant is a resource or POJO that participates in an LRA. It contains annotations for managing callbacks from the narayana LRA coordinator (via a proxy in the non JAX-RS case). Minimally, an LRA participant must contain a method annotated with @Compensate which is the callback to use when the LRA is cancelled and it is used to inform the participant that it should compensate for any actions it performed when the method marked with @LRA was invoked. This callback is guaranteed to be eventually delivered. And, optionally, if a method is annotated with @Complete then the participant will be notified if the LRA is closed.

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+When such a method is called a context is created (if one is not already present) which is passed
+along with subsequent JAX-RS invocations until a method is reached
+which also contains an `@LRA` annotation with an attribute that indicates that the LRA should be
+closed or cancelled. The default is for the LRA to be closed in the same method that started the
```
Is close a successful termination, while cancel is an unsuccessful one?

mmusgrov replied:
The state model is shown in [Figure 2 of the spec where we define the state model](https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc#the-model). Would you prefer that I define them here in the guide?

mmusgrov replied:
But yes, in JTA terminology, close is the equivalent of `commit` and cancel is the equivalent of `rollback`.
The terms used in the model come from the original [Oasis WS-LRA spec](https://www.oasis-open.org/committees/document.php?document_id=12794).

cescoffier replied:
I believe this should be written somewhere, maybe using an admonition. Again, readers may not be familiar with the LRA spec (and won't read it).

mmusgrov replied:
ok, I already say that when the context is cancelled the compensate method is called, but I can spell it out.

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+ran to completion).
+
+Note that this version of the extension only supports JAX-RS based services (and a subsequent
+revision will include support for non JAX-RS based services).
```
Ok, got my answer. But isn't it an interceptor?

mmusgrov replied:
The narayana implementation requires that the application supports JAX-RS but it does not require the participants to be JAX-RS resources. We have a CDI extension that parses the application deployment and creates proxies for interaction with our coordinator.

Before I took the implementation to the spec group I wrapped then non JAX-RS support behind an interface. The application user was then responsible for starting, joining with and for ending LRAs. The new model uses annotations to take such a burden away from the application but we still use similar techniques.

cescoffier replied:
Thanks!

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+new LRAs and to enlist participants with them.
+
+The LRA extension can be configured by placing an `application.properties` file
+in the `src/main/resources` directory. The only LRA specific property is
```
This file is automatically created when you create the Quarkus project.

mmusgrov replied:
I could reword it as `by updating the` instead of `by placing an`. I have never used any tool to `create the Quarkus project`, I just write the pom like I would write any java application. Do you have a feel for how many people use the tool you are referring to?

cescoffier replied:
Most users use code.quarkus.io. Users writing the pom.xml themselves are generally more advanced and so now when (and where) to create the file. So, +1 to reword as you proposed.

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+or you can build your own coordinator using a maven pom that includes the appropriate
+dependencies. A quarkus quickstart will be provided to show how to do this or you can
+take a look at one of the https://github.com/jbosstm/quickstart/tree/master/rts/lra-examples/lra-coordinator[narayana quickstarts].
+Another option would be to run it managed inside a WildFly application server.
```
Oh... that lead to dev services :-D 

We may want to automatically start a coordinator in dev/test modes.

mmusgrov replied:
How do I do that. In the integration test I use @QuarkusTestResource to hook into the test lifecycle and I start a coordinator in a docker container?

cescoffier replied:
Dev Services is kind of the same thing but run during test and dev mode. We can do that in another PR (a few things are still in flux around dev services). Can you open an issue so we don't forget?

mmusgrov replied:
#18425

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+    @LRA(LRA.Type.REQUIRES_NEW)
+    @Path("/work")
+    @PUT
+    public void doInTransaction(@HeaderParam(LRA_HTTP_CONTEXT_HEADER) URI lraId)
```
I know that transaction and LRA are close, but, it would be better to avoid any potential confusion.

mmusgrov replied:
An LRA is a transaction, although not full ACID because we relax C and I, but it is still a transaction.

mmusgrov replied:
You have provided a lot of useful comments, many of which require a little bit of dev work so I will go through all those first. When I've done that I will provide responses/resolutions to you comments.

For this particular comment, I will rename it `doInLongRunningAction` and include a comment that makes it clear that an LRA is not full ACID.

cescoffier replied:
+1

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+         * value in the injected JAX-RS header. This LRA was started just before
+         * the method was entered (REQUIRES_NEW) and will be closed when the
+         * method finishes at which point the completeWork method below will be
+         * invoked.
```
Again for my own understanding. 
You get the request and intercept it. You extract the header and create/enlist the participant into the coordinator. Then, depending on the completion, you notify the coordinator. Am I right?

mmusgrov replied:
The JAX-RS request filter intercepts incoming JAX-RS requests to see if the target resource contains LRA annotations. If the resource is an LRA participant then it interacts with the coordinator (whether it's remote or in-VM) in order to manage the life-cycle and membership of the LRA. If the coordinator is unavailable it reports the failure back to the caller. The filter interacts with the coordinator using timed async JAX-RS invocations ie it is effectively a blocking call (at the time I wrote it there was no support for async filters and there probably still isn't, although Stephane (Epardaud) did write one that I briefly experimented with).

cescoffier replied:
Thanks

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+         */
+    }
+
+    @Complete
```
Probably better to use FQN, as it's not clear that this is an LRA annotation.

mmusgrov replied:
okay

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+         */
+         return Response.ok(ParticipantStatus.Compensated.name()).build();
+    }
+}
```
@Complete and @Compensate are per participant, right?

mmusgrov replied:
Yes (the details are in the spec, if you want the details I will find the exact reference).

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+
+The same section of the specification document also includes an
+example that demonstrates how to start an LRA in one method and close
+it in a different resource method using the `LRA#end` element.
```
`LRA.end` ?

mmusgrov replied:
okay

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+The example also shows that when an LRA is present its identifier can be obtained
+by reading the request headers.
+
+The same section of the specification document also includes an
```
I found it rather odd to refer to the spec. I would not expect our users to read it (they won't :-)). 
So, I would rephrase to provide examples.

mmusgrov replied:
okay, although isn't it still worth mentioning that the spec and TCK contain many other examples?

cescoffier replied:
Yes, definitely mention it, but your narrative should not expect readers to look at it.

mmusgrov replied:
Nowhere does the narrative require readers to read the spec. The narrative only said that the example was taken from the spec.

cescoffier commented on extensions/narayana-lra/deployment/pom.xml:
```diff
@@ -0,0 +1,54 @@
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-jsonb-deployment</artifactId>
```
Is there any dependency on JSON-B? Quarkus recommends Jackson.

mmusgrov replied:
No. I did tryt modifying the dependencies and got errors like:
```
Caused by: java.lang.IllegalStateException: Please make sure there is only one provider of the following capabilities:
capability io.quarkus.rest is provided by:
  - io.quarkus:quarkus-resteasy-reactive::jar:999-SNAPSHOT
  - io.quarkus:quarkus-resteasy::jar:999-SNAPSHOT
```
I couldn't figure who was still pulling the non reactive version but I will investigate further.

cescoffier commented on extensions/narayana-lra/deployment/pom.xml:
```diff
@@ -0,0 +1,54 @@
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-rest-client-deployment</artifactId>
```
Did you try with the reactive rest client (even used in a blocking manner)? It would be better to use the reactive one as it's more integrated with the Quarkus internals. 

CC @michalszynkiewicz

michalszynkiewicz replied:
I think the client is by [this class](https://github.com/jbosstm/narayana/blob/master/rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java). If it is so, the async api seems to be the only one in use.

mmusgrov replied:
I will give it a go

mmusgrov replied:
Yes @michael-schnell we only use the async API. Does that mean I don't need to use the reactive rest client.

NB When I switch to using `<artifactId>quarkus-rest-client-reactive-deployment</artifactId>` I get the following runtime error:
            `ClassNotFoundException: io.quarkus.resteasy.reactive.common.runtime.ResteasyReactiveConfig`
Is is worth debugging this given that we are already using the async version of the API?

cescoffier commented on extensions/narayana-lra/runtime/pom.xml:
```diff
@@ -0,0 +1,66 @@
+        <dependency>
+            <groupId>org.jboss.narayana.rts</groupId>
+            <artifactId>lra-proxy-api</artifactId>
+            <scope>compile</scope>
```
Any reason to set the scope?

cescoffier replied:
Yes, probably better.

cescoffier commented on extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/LRAConfiguration.java:
```diff
@@ -0,0 +1,23 @@
+     * In this version of the extension, a failed coordinator with
+     * LRAs that have not yet finished must be restarted.
+     */
+    @ConfigItem(defaultValue = "http://localhost:8080/lra-coordinator")
```
8080 is the default port of Quarkus. So except if you expect the Quarkus to be also the coordinator, it would need to be changed.

cescoffier replied:
I'm a bit surprised by the configuration, what about TLS, certs and so on.

mmusgrov replied:
I will change the port to one of the PrivatePorts (49152-65535).

mmusgrov replied:
How do I test TLS, certs and so on, I don't have the expertise to do that, is there someone in the quarkus team that can make an additional PR once this current one has been merged?

cescoffier replied:
Let's do this in an additional PR.

cescoffier commented on integration-tests/narayana-lra/pom.xml:
```diff
@@ -0,0 +1,169 @@
+        </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-resteasy-deployment</artifactId>
```
I would test with classic resteasy but also with resteasy reactive. Should not change anything, but we never know.

mmusgrov replied:
I will give it another go. I did try modifying the dependencies and got errors like:
```
Caused by: java.lang.IllegalStateException: Please make sure there is only one provider of the following capabilities:
capability io.quarkus.rest is provided by:
  - io.quarkus:quarkus-resteasy-reactive::jar:999-SNAPSHOT
  - io.quarkus:quarkus-resteasy::jar:999-SNAPSHOT
```
I couldn't figure who was still pulling the non reactive version but I will investigate further.

mmusgrov replied:
See my earlier comment above (https://github.com/quarkusio/quarkus/pull/17903#discussion_r663043617)

cescoffier replied:
You need to remove `io.quarkus:quarkus-resteasy`, both can't be used at the same time.

mmusgrov replied:
That is what I did.

cescoffier replied:
Hum, that's kind of annoying. Does your implementation pull the classic Resteasy. We would need to have a look.

cescoffier commented on integration-tests/narayana-lra/src/main/resources/application.properties:
```diff
@@ -0,0 +1,9 @@
+quarkus.http.test-port=8081
+quarkus.http.host=${HOST:localhost}
+quarkus.ssl.native=true
```
Should we automatically enabled SSL in the LRA extension?

mmusgrov replied:
I will remove it.

cescoffier commented on integration-tests/narayana-lra/src/test/java/org/acme/quickstart/lra/LRAParticipantTest.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    // send an HTTP GET request to an endpoint
+    private static String coordinatorGetRequest(String endpoint, int[] expectedResponseCodes, boolean readResponse)
```
Why not using restassured for this? You can pass absolute URLs too.

mmusgrov replied:
The routine uses Jupiter:
```java
assertTrue(Arrays.stream(expectedResponseCodes).anyMatch(i -> i == responseCode),
                "unexpected response code: GET " + endpoint);
```

cescoffier reviewed (changes requested):
Lots of comments are actually questions to satisfy my own curiosity.

mmusgrov commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+The idea is for multiple services to perform different computations/actions in concert, whilst
+retaining the option to compensate for any actions performed during the computation.
+This kind of loose coupling of services bridges the gap between strong consistency models
+such as JTA/XA and "home grown" ad hoc consistency solutions.
```
The standard dictionaries use ad hoc. Though I could remove ad hoc and the sentence will still read correctly. What do you prefer?

cescoffier replied:
ad hoc is fine, we wrote it with a dash.

mmusgrov commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+The model is based on the https://github.com/eclipse/microprofile-lra/blob/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc#eclipse-microprofile-lra[Eclipse Microprofile LRA specification].
+The approach is for the developer to annotate a business method with a Java annotation
+(https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java[`@LRA`]).
+When such a method is called a context is created (if one is not already present) which is passed
```
+!

mmusgrov commented:
@cescoffier I responded to your comments but I won't push changes until I know you are happy with how I plan to address those comments that require updates to the extension. So if you wouldn't mind, can you confirm whether or not my responses are adequate?

cescoffier commented on docs/src/main/asciidoc/lra-guide.adoc:
```diff
@@ -0,0 +1,196 @@
+The LRA (short for Long Running Action) participant extension is useful in microservice based
+designs where different services can benefit from a relaxed notion of distributed consistency.
+
+The idea is for multiple services to perform different computations/actions in concert, whilst
```
It just making reviews and suggestions a lot easier.

mmusgrov replied:
Perhaps, but it is not the correct use of the English language, paragraphs are there to group of related ideas to facilitate comprehension, the guide is provide for developers, not reviewers.

cescoffier replied:
Asciidoc does not work like that. New lines do not introduce paragraphs. 

For example:
```
This is my first sentence.
This is a second sentence in the same paragraph.

This is another paragraph.
```

mmusgrov replied:
I don't follow your point, if you generate html from the asciidoc (`asciidoc lra-guide.adoc`) you will find that the text reads fine and paragraphs are displayed correctly.

mmusgrov commented:
@cescoffier From my viewpoint it seems that I have resolved each of your objections, ie I don't plan to push any more updates unless you have new input. Can you confirm whether or not you are waiting for me to provide new input?

mmusgrov commented:
@NahuelOlgiati and @xstefank  I updated this PR adding non JAX-RS support.

I included a hack to make the constructors for io.narayana.lra.client.internal.proxy.nonjaxrs.LRAParticipant and LRAParticipantRegistry public. I tried using reflection but could not get it working. I decided not to spend too much time on that since we may (or may not) need to rework the narayana support for non JAX-RS so that it doesn't use CDI extensions.

I will add some more integration tests when I return from my PTO (but my quickstart works).

gsmet commented:
Just a note that I haven't forgotten about this, it's slowly moving to the top of my list while I go through my backlog and my usual daily tasks.

gsmet commented:
@mmusgrov can you give me a window when I can work on this PR? It needs a rebase (we don't want merge commits in PRs) and I might have to make a couple of small adjustments here and there. Would Friday work for you?

mmusgrov commented:
@gsmet Friday works okay for me.
I used the github merge tool for resolving the conflict, I will force push another update ...

mmusgrov commented:
@gsmet I think the PR is in good shape but I have some improvements to the manual that could be of help for developers. May I update it or should I wait?

gsmet commented on extensions/narayana-lra/runtime/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipant.java:
```diff
@@ -1,24 +1,3 @@
- * License along with this software; if not, write to the Free
- * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
- * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
- */
```
So... these files are a problem for me:
- first, Quarkus is Apache licensed so we need to relicense them under the right license. I have no idea where they come from so I'll let you figure out if it's doable or not;
- second are they somehow overriding files in existing jars? It looks like it from what I can see in the other jars. That's definitely something we don't want to do: we don't want a split package and we don't want to override classes this way. I have no idea why you did that so we would need to understand the reason to help you solve this problem.

mmusgrov replied:
These files were copied over from the narayana implementation of LRA, I needed access to the constructor which is package private. I have changed them in the narayana repo (https://github.com/jbosstm/narayana/pull/1888) so when we next do a narayana release I will update this extension and remove the two files (LRAParticipant.java and LRAParticipantRegistry.java).

gsmet commented on extensions/narayana-lra/runtime/src/main/java/io/quarkus/narayana/lra/runtime/NarayanaLRAProducers.java:
```diff
@@ -16,7 +15,6 @@ public LRAParticipantRegistry lraParticipantRegistry() {
 
     @Produces
     public ParticipantProxyResource participantProxyResource() {
-        Object proxy = CDI.current().select(ParticipantProxyResource.class);
```
This variable is unused. Not sure why you did that or if it's just a leftover?

mmusgrov replied:
Thanks

mmusgrov commented:
@gsmet The changes look good

gsmet commented on extensions/narayana-lra/runtime/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRAParticipantRegistry.java:
```diff
@@ -1,24 +1,3 @@
- * License along with this software; if not, write to the Free
- * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
- * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
- */
```
Same remark here.

mmusgrov replied:
Thanks (see my remark).

gsmet reviewed (changes requested):
@mmusgrov I pushed a commit with some small changes.

There's one blocking problem though: we need to figure out what to do with the classes in `io.narayana.lra.client.internal.proxy.nonjaxrs` (see my comment inline). Maybe ping me on Zulip to discuss it?

mmusgrov commented:
@gsmet I'll create a blog at `quarkus.io/blog` corresponding to the release of the artefact. Can I rely on the extension being listed in the catalog or should tell readers to manually add the dependency `io.quarkus:quarkus-narayana-lra` to their poms and if so what version should I tell them to use. And who should I submit my blog to?

gsmet commented:
Squashed and rebased. Let's wait for CI and merge.

gsmet commented:
@mmusgrov when it's merged, it will get available when 2.2 is released. Schedule is here: https://github.com/quarkusio/quarkus/wiki/Release-Planning .

If everything is fine, it should appear automatically in code.quarkus.io, yes.

mmusgrov commented:
@gsmet `COMMITTERS.adoc` says that if the feature is worth mentioning in the release notes and announcement blog post then the PR requires the `release/noteworthy-feature` label. Will you add that for me?

@gsmet Also I've written a blog to introduce the extension in adoc format, who should I give it to (there's no repo for blogs) so that it appears at https://quarkus.io/blog?

gsmet commented:
Merged and added the `release/noteworthy-feature`.

As for the blog posts, they are there: https://github.com/quarkusio/quarkusio.github.io/tree/develop/_posts
And you need to add your author information here: https://github.com/quarkusio/quarkusio.github.io/blob/develop/_data/authors.yaml

mmusgrov commented:
Thanks @gsmet - I have have added a [PR for the blog](https://github.com/quarkusio/quarkusio.github.io/pull/1105). Note that I have just returned from PTO so apologies for its late submission. Is there anyone I should ping to get it added/reviewed?
